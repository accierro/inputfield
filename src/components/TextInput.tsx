import * as React from "react";
import styled from "styled-components";
import { IoIosWarning } from "react-icons/io";

import { colors } from "../colorConfig";
import StyledTextInput from "../styled/StyledTextInput";
import IconWithMessage from "./IconWithMessage";

type TextInputProps = {
  label: string;
  optional?: boolean;
  value: string;
  error?: string;
  onChange: (s: string) => void;
};

const OptionalLabel = styled.span`
  font-size: 14px;
  font-weight: 300;
  color: ${colors.mediumGrey};
  margin-left: 8px;
`;

type LabelProps = {
  error: boolean;
};

const Label = styled.label`
  font-size: 18px;
  font-weight: normal;
  margin-bottom: 4px;
  display: flex;
  align-items: center;
  color: ${(props: LabelProps) =>
    props.error ? colors.error : colors.darkGrey};
`;

const TextInput: React.FC<TextInputProps> = ({
  label,
  optional = false,
  value,
  error,
  onChange,
}) => {
  // const [internalError, setInternalError] = React.useState<string | null>(null);

  const onChangeHandler = React.useCallback(
    (e) => {
      onChange(e.target.value);
    },
    [onChange]
  );
  // const onBlurHandler = React.useCallback(
  //   (e) => {
  //     if (validationFn) {
  //       const result = validationFn(e.target.value);
  //       if (result.error) {
  //         setInternalError(result.message ?? "Validation error");
  //       } else {
  //         setInternalError(null);
  //       }
  //     }
  //   },
  //   [validationFn]
  // );
  return (
    <div
      style={{
        display: "flex",
        flexDirection: "column-reverse",
      }}
    >
      <StyledTextInput
        error={error !== undefined}
        type="text"
        value={value}
        onChange={onChangeHandler}
      />
      <Label error={error !== undefined}>
        {label}
        {optional && <OptionalLabel>Optional</OptionalLabel>}
        {error && (
          <IconWithMessage
            icon={
              <IoIosWarning
                size={20}
                color={colors.error}
                style={{ marginLeft: "8px" }}
              />
            }
            message={error}
          />
        )}
      </Label>
    </div>
  );
};

export default TextInput;
