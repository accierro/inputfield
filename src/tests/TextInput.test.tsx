import * as React from "react";
import TextInput from "../components/TextInput";
//@ts-ignore
import {
  render,
  cleanup,
  fireEvent,
  getByText,
  getByRole,
  queryByText,
} from "@testing-library/react";

describe("<TextInput />", () => {
  afterEach(() => {
    cleanup();
  });
  test("matches a snapshot", () => {
    const { container } = render(
      <TextInput label="Test" value="" onChange={() => {}} />
    );
    expect(container).toMatchSnapshot();
  });

  test("shows optional if the prop is passed", () => {
    const { container, rerender } = render(
      <TextInput label="Test" optional value="" onChange={() => {}} />
    );
    expect(getByText(container, "Optional")).toBeInTheDocument();

    rerender(<TextInput label="Test" value="" onChange={() => {}} />);
    expect(queryByText(container, "Optional")).toBeNull();
  });

  test("shows warning error if error prop is passed", () => {
    const { container } = render(
      <TextInput
        label="Test"
        value=""
        error="Can't be empty"
        onChange={() => {}}
      />
    );
    const svg = container.querySelector("svg");
    expect(svg).toBeInTheDocument();

    if (svg) {
      fireEvent.mouseOver(svg);
    }
    expect(getByText(container, "Can't be empty")).toBeInTheDocument();
  });

  test("displays correct value", () => {
    const { container, rerender } = render(
      <TextInput label="Test" value="Hello" onChange={() => {}} />
    );

    expect(getByRole(container, "textbox").getAttribute("value")).toEqual(
      "Hello"
    );

    rerender(<TextInput label="Test" value="1" onChange={() => {}} />);
    expect(getByRole(container, "textbox").getAttribute("value")).toEqual("1");
  });

  test("onChange event is fired", () => {
    const mockFn = jest.fn();
    const { container } = render(
      <TextInput label="Test" value="Hello" onChange={mockFn} />
    );
    fireEvent.change(getByRole(container, "textbox"), {
      target: { value: "check" },
    });
    expect(mockFn).toBeCalledTimes(1);
    expect(mockFn).toBeCalledWith("check");
  });
});
