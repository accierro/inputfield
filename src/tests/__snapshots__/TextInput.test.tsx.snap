// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> matches a snapshot 1`] = `
<div>
  <div
    style="display: flex; flex-direction: column-reverse;"
  >
    <input
      class="sc-AxjAm epslib"
      type="text"
      value=""
    />
    <label
      class="sc-AxiKw bpTyXb"
    >
      Test
    </label>
  </div>
</div>
`;
